droidmediabuffer.cpp

```diff
@@ -138,31 +159,83 @@ DroidMediaBuffer *droid_media_buffer_create_from_raw_data(uint32_t w, uint32_t h
     return NULL;
   }
 
-  dstStride = buffer->getStride();
+  linesY = h;
+  dstStrideY = buffer->getStride();
+
+  switch (format) {
+  case HAL_PIXEL_FORMAT_RGBA_8888:
+  case HAL_PIXEL_FORMAT_RGBX_8888:
+  case HAL_PIXEL_FORMAT_BGRA_8888:
+    lineWidthY = 4 * w;
+    break;
+  case HAL_PIXEL_FORMAT_RGB_888:
+    lineWidthY = 3 * w;
+    break;
+  case HAL_PIXEL_FORMAT_RGB_565:
+  case HAL_PIXEL_FORMAT_YCbCr_422_I:
+    lineWidthY = 2 * w;
+    break;
+  case HAL_PIXEL_FORMAT_YV12:
+  case HAL_PIXEL_FORMAT_YCbCr_422_SP:
+  case HAL_PIXEL_FORMAT_YCrCb_420_SP:
+  case /* QOMX_COLOR_FormatYUV420PackedSemiPlanar32m */ 0x7FA30C04:
+  case /* QOMX_COLOR_FormatYUV420PackedSemiPlanar64x32Tile2m8ka */ 0x7FA30C03:
+    lineWidthY = w;
+    break;
+  }
 
-  if (strideY == dstStride) {
+  switch (format) {
+  case HAL_PIXEL_FORMAT_YV12:
+    dstStrideUV = dstStrideY / 2;
+    // U plane is followed by V plane. So let's do 2-in-1 by not halving h.
+    linesUV = h;
+    lineWidthUV = w / 2;
+    break;
+  case HAL_PIXEL_FORMAT_YCbCr_422_SP:
+    linesUV = h; // Twice the lines as 4:2:0.
+    lineWidthUV = w;
+    dstStrideUV = dstStrideY;
+    break;
+  case HAL_PIXEL_FORMAT_YCrCb_420_SP:
+  case /* QOMX_COLOR_FormatYUV420PackedSemiPlanar32m */ 0x7FA30C04:
+  case /* QOMX_COLOR_FormatYUV420PackedSemiPlanar64x32Tile2m8ka */ 0x7FA30C03:
+    linesUV = h / 2;
+    lineWidthUV = w;
+    dstStrideUV = dstStrideY;
+    break;
+  default: // RGB, and YCbCr_422_I
+    if (strideUV != 0)
+      ALOGW("Get unexpected 2nd plane for the format that shouldn't have one.");
```

Comment:
In this case, the global memcpy below should still happen if the caller erroneously passed a strideUV for an RGB format.

Reply:
I've made it set the `strideUV` to 0 in this case. If `strideY` equals to `dstStrideY`, the global `memcpy` should still happen.